src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1391,19 @@ struct OptimizeInstructions
       right->type = Type::i32;
       return right;
     }
+    {
+      Expression* x;
+      // ~(1 << x) aka (1 << x) ^ -1  ==>  rotl(-2, x)
+      if (matches(curr,
+                  binary(Abstract::Xor,
+                         binary(&curr, Abstract::Shl, ival(1), any(&x)),
+                         ival(-1)))) {
+        curr->op = Abstract::getBinary(type, Abstract::RotL);
+        curr->left->cast<Const>()->value = Literal::makeFromInt32(-2, type);
```

Comment:
This doesn't seem right since `curr->left` is a binary op, not a constant.

Edit: Oh, I see, you're reassigning `curr` in the match. Unfortunately that's undefined behavior because it will only work if the arguments to `matches` are evaluated left-to-right, which is not guaranteed.

```suggestion
        right->value = Literal::makeFromInt32(-2, type);
        curr->left = right;
```

Reply:
> Oh, I see, you're reassigning curr in the match

It's your suggestion btw. Perhaps better revert it to different variable for more obvious view?

Reply:
> Unfortunately that's undefined behavior because it will only work if the arguments to matches are evaluated left-to-right.

Hmm, what does it mean? I thought `1 << x` will be always has same order due to `x << 1` or `x >> 1` will be do absolutely different things and can't happening

Reply:
Ah sorry, I got myself very confused there. There's no problem after all.

Reply:
Proposed way also works great) Just little bit more tricky